### PR TITLE
[bugfix] deck.gl grid&hex don't respond to zoom&drag

### PR DESCRIPTION
**Pull request by mistercrunch**
Following recent changes, Hex & Grid component wouldn't respond to
viewport interactions. This fixes the issue.

This PR also greatly improve rendering performance by only recomputing
the deck.gl layers when needed, as opposed to every time the viewport
changes, which is very inneficient.

Note that most other deck.gl layers seem to be systematically
recomputing layers, we'll have to fix this in another PR.

### CODE REVIEW

**Comment by mistercrunch** [comment]
FYI this is running in production at Lyft since Nov 12th

File: superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx [hunk 1]
```diff
@@ -23,12 +24,13 @@ const defaultProps = {
 export default class DeckGLContainer extends React.Component {
   constructor(props) {
     super(props);
+    this.tick = this.tick.bind(this);
+    this.onViewportChange = this.onViewportChange.bind(this);
```
Comment (hughhhh) [inline superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx]: nit: use autobind here
https://github.com/andreypopp/autobind-decorator

File: superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx [hunk 2]
```diff
@@ -53,7 +55,9 @@ export default class DeckGLContainer extends React.Component {
   }
   tick() {
     // Limiting updating viewport controls through Redux at most 1*sec
-    if (this.state && this.state.previousViewport !== this.props.viewport) {
+    // Deep compare is needed as shallow equality doesn't work here, viewport object
+    // changes id at every change
```
Comment (hughhhh) [inline superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx]: nit:
```
const { previousViewPort } = this.state;
const { viewport } = this.props;
if (this.state && !isEqual(previousViewport, viewport)) {
   ...
}
```

File: superset/assets/src/visualizations/deckgl/factory.jsx [hunk 3]
```diff
@@ -18,36 +20,62 @@ const defaultProps = {
...
+      super(props);
+      const originalViewport = props.viewport;
+      const viewport = props.formData.autozoom
+        ? fitViewport(originalViewport, getPoints(props.payload.data.features))
+        : originalViewport;
+      this.state = {
+        viewport,
+        layer: this.computeLayer(props),
+      };
+      this.onViewportChange = this.onViewportChange.bind(this);
```
Comment (hughhhh) [inline superset/assets/src/visualizations/deckgl/factory.jsx]: nit: use autobind
https://github.com/andreypopp/autobind-decorator

File: superset/assets/src/visualizations/deckgl/factory.jsx [hunk 4]
```diff
@@ -18,36 +20,62 @@ const defaultProps = {
...
+    render() {
+      const {
+        formData,
+        payload,
+        setControlValue,
+      } = this.props;
+      return (
+        <DeckGLContainer
+          mapboxApiAccessToken={payload.data.mapboxApiKey}
+          viewport={this.state.viewport}
```
Comment (hughhhh) [inline superset/assets/src/visualizations/deckgl/factory.jsx]: deconstruct `this.state`
```
const {
   layer,
   viewport
} = this.state
```

File: superset/assets/src/visualizations/deckgl/factory.jsx [hunk 5]
```diff
@@ -18,36 +20,62 @@ const defaultProps = {
...
-        layers={[layer]}
-        mapStyle={formData.mapbox_style}
-        setControlValue={setControlValue}
-      />
-    );
+  // Higher order component
+  class Component extends React.PureComponent {
+    constructor(props) {
+      super(props);
+      const originalViewport = props.viewport;
```
Comment (hughhhh) [inline superset/assets/src/visualizations/deckgl/factory.jsx]: deconstruct props

File: superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx (hunk 1 above)
Comment (mistercrunch) [reply superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx]: We should introduce `autobind` all at once, would rather not have it only in some places.

File: superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx (hunk 2 above)
Comment (mistercrunch) [reply superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx]: Careful here though `const { previousViewPort } = this.state;` will fail if `this.state` is `undefined`

Comment (hughhhh) [reply superset/assets/src/visualizations/deckgl/DeckGLContainer.jsx]: isn't the state always going to be defined with this line:
https://github.com/apache/incubator-superset/pull/6363/files/052fe92b28888a5d428575a0dde50bd4a7900388#diff-52ae534f640f8c5b7eb6e05bd0e56259R30

**Comment by mistercrunch** [comment]
Ok addressed most comments, saving @autobind for a later cross codebase PR.

**Pull request merged**